### ISSUE TITLE
SQLSTATE[HY000]: General error: 1215 Cannot add foreign key constraint

### ISSUE THREAD

**Issue by markxxv** [rokorolov/parus#2]
Hi! When i trying to run last migration

`php yii migrate --migrationPath=@rokorolov/parus/gallery/migrations`

have that error in terminal:

`Exception 'yii\db\Exception' with message 'SQLSTATE[HY000]: General error: 1215 Cannot add foreign key constraint
The SQL being executed was: ALTER TABLE `album_lang` ADD CONSTRAINT `fk__album_lang_language__language_lang_code` FOREIGN KEY (`language`) REFERENCES `language` (`lang_code`) ON DELETE CASCADE ON UPDATE CASCADE'

in /Users/mark/parus/vendor/yiisoft/yii2/db/Schema.php:631

Error Info:
Array
(
    [0] => HY000
    [1] => 1215
    [2] => Cannot add foreign key constraint
)

`
I tried to run this migration on different new DB but every time had same error. How can i solve it? 

p.s thanx for this awesome app its a perfect base for any new site on Yii!

**Comment by rokorolov** [comment]
Hi, what's your MySQL version?

**Comment by markxxv** [comment]
My version is 

`mysql  Ver 14.14 Distrib 5.7.14, for osx10.11 (x86_64) `

Thanks for any idea

**Comment by rokorolov** [comment]
Yes, there can be a problem because was not testing on mysql 5.7. Now I need to install new version and do some tests. But it will take some time.

**Comment by markxxv** [comment]
Thank you! 

And what version are you using?

**Comment by roxblnfk** [comment]
in mysql 5.6 same error
m160914_124800_init_module_gallery.php [line 29]
m160914_124800_init_module_gallery.php [line 56]
I just commented out these lines :|

**Comment by rokorolov** [comment]
One more issue on mysql 5.7.16 with datetime default value '0000-00-00 00:00:00' [no_zero_date](http://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_no_zero_date)

> And what version are you using?

I was only testing on MariaDB and MySQL 5.5, now testing on the latest mysql version. And will update as soon as possible.

**Comment by rokorolov** [comment]
@markxxv and @roxblnfk need your help (can't get the described foreign key error on mysql 5.7.16 but get another). Can you please test the new migrations. Just install beta version (change to `"rokorolov/parus": "@dev"` in `composer.json`), and run in console `php yii migrate --migrationPath=@rokorolov/parus/admin/migrations`. In the new version all migrations in one place.

Can't repair by adding new migration so I rewrote them while the project is under alpha stage. It's not a good practice but I don't have another choice.

**Comment by roxblnfk** [comment]
It`s works but needed to make a fix:

config/web.php[92]:
```
 'db' => include_once __DIR__ . '/db.php',
```
config/db.php:
```
<?php
return array_merge(
	[
		'class' => 'yii\db\Connection',
		'dsn' => 'mysql:host=localhost;dbname=yii2basic',
		'username' => 'root',
		'password' => '',
		'charset' => 'utf8',
	],
	is_file(__DIR__ . DIRECTORY_SEPARATOR . 'db-local.php') ? include_once __DIR__ . DIRECTORY_SEPARATOR . 'db-local.php' : []
);
```

And if run in console **yii run** and **yes**:

	Start applying Migrations ...
	Exception 'yii\base\InvalidConfigException' with message 'Migration failed.
	Directory specified in migrationPath doesn't exist: @rokorolov/parus/language/migrations'
	in \parus\vendor\yiisoft\yii2\console\controllers\BaseMigrateController.php:103

**Comment by rokorolov** [comment]
> but needed to make a fix:

You made these changes :), so pull request and i will merge with master branch.

> And if run in console yii run and yes:

Yes, this happens because you have stable version without last changes, all changes on beta version now `composer create-project --prefer-dist rokorolov/parus-basic-app parus-basic @dev`

Tomorrow I will apply all this changes to stable version.

**Comment by markxxv** [comment]
I tried with `composer create-project --prefer-dist rokorolov/parus-basic-app parus-basic @dev`
but i have the same issue 

```
Exception 'yii\base\InvalidConfigException' with message 'Migration failed. Directory specified in migrationPath doesn't exist: @rokorolov/parus/admin/migrations'
```

and if i trying one to one, i have same issue with this `1067 Invalid default value for 'created_at'`

**Comment by rokorolov** [comment]
You need specify dev version and for module, change to "rokorolov/parus": "@dev" in composer.json and update with composer.

**Comment by markxxv** [comment]
Yes, now it works now, thank you much. I've updated app with composer on @dev. Then migrate like this: ```php yii migrate --migrationPath=@rokorolov/parus/admin/migrations```

Migrations was successful. Admin panel work perfect.

So i have another new issue, it’s about menu and 404 error. Should I create new issue or continue right here? (sorry i'm new on git)

**Comment by rokorolov** [comment]
>  it’s about menu and 404 error

Try to clean cache first and if that doesn't helps, create new issue.

**Comment by tararaden** [comment]
Такая же проблема была. Оказывается, не получалось присвоить внешний ключ из-за разных Сравнений в полях таблиц. Почему-то в одних миграциях (напр. Меню) сравнение устанавливается при создании таблицы, а в других (напр. Фотоальбом) нет. Может нужно либо везде ставить, либо нигде не ставить.

**Comment by jkmssoft** [comment]
My mysql default character set is latin1_swedish_ci.
But you are using utf8_unicode_ci.
You forgot to set the caracter in the gallery migration file.
A foreign key from character set latin1_swedish_ci to a table with utf8_unicode_ci doesn't work.

Solution: I recommend to create a own Migration class which does set the tableOptions in the init()
method and then using this class as the parent one in all migrations.
Then use `$this->tableOptions` in all migrations and createTable commands.

```php
<?php

namespace rokorolov\parus\admin\migrations;

/**
 * @author jkmssoft
 */
class Migration extends \yii\db\Migration
{
    /**
     * @var string
     */
    protected $tableOptions;

    /**
     * @inheritdoc
     */
    public function init()
    {
        parent::init();

        if ($this->db->driverName === 'mysql') {
            $this->tableOptions = 'CHARACTER SET utf8 COLLATE utf8_unicode_ci ENGINE=InnoDB';
        }
    }
}
```

**Comment by rokorolov** [comment]
That's where the bug, thanks @jkmssoft
